### ISSUE TITLE
[h2] Deferred reset stream does not work when h2 session window is full

### ISSUE THREAD

**Issue by ppadevski-vmw** [envoyproxy/envoy#29489]
I'm using v1.26.3 though we've been seeing this issue for quite some time.

I have the following use case:
client =>h2=> envoy =>h1=> server (all localhost)

The client seems to get RST_STREAMs occasionally though it is very difficult to reproduce.

The pcaps suggest the following scenario:

client sends 65kb packet that has multiple requests (15+) in a single packet
```
...
HyperText Transfer Protocol 2
    Stream: HEADERS, Stream ID: 3425, Length 62, POST /URL
...
HyperText Transfer Protocol 2
    Stream: DATA, Stream ID: 3425, Length 3410 (partial entity body)
...
```

So far so good. The above supposedly contains the entire request (headers and body).

At this point the request is forwarded to the server via h1 and the server replies in less than 1ms.

The request contains Content-Length.

The response is seemingly returned in 1 packet around 1.7kb (localhost) which contains both headers and body.

At this point we see envoy send the RST_STREAM seemingly due to:

```
void Filter::onUpstreamComplete(UpstreamRequest& upstream_request) {
  if (!downstream_end_stream_) {
    upstream_request.resetStream(); // <<<<<<<<<<<<<<<<<<<<<<<<<<<
  }
```

and at this point we can notice that ~1ms LATER (or at the exact same time) the pcap contains:

```
Frame 19821: 7131 bytes on wire (57048 bits), 7131 bytes captured (57048 bits)
...
HyperText Transfer Protocol 2
    Stream: DATA, Stream ID: 3425, Length 0
```

which contains no headers and body for Stream ID: 3425 and just h2 End Stream. Maybe this is why downstream_end_stream_ is false.

The odd part is that when this is hit usually there are ~5-10 responses that fail with missing body and some that somehow work, maybe nghttp2 drops some data after it sees the RST_STREAM but it does not seem very deterministic to me.

Now, I was told that the empty End Stream a few ms later is according to RFC though I haven't verified this myself.

At this point it seems that I can't put the blame on either the client, the server or envoy but I need to find a solution and h2 for the server would be very painful. I cannot say whether getting rid of Content-Length would make any difference.

To summarize:
1. h2: ===> HEADERS, BODY (around 4k)
2. h1: ===> request (sent in 4 packet, ~4k)
3. h1: <=== response (less than 1ms later, received in 1 packet, around 1.7k)
4. h2: <=== RST_STREAM
5. h2: ===> End Stream (less than 1ms after (4) or at the exact same time)

**Comment by ppadevski-vmw** [comment]
Actually according per RFC:

https://httpwg.org/specs/rfc9113.html#FlowControl

`"A server can send a complete response prior to the client sending an entire request if the response does not depend on any portion of the request that has not been sent and received. When this is true, a server MAY request that the client abort transmission of a request without error by sending a RST_STREAM with an error code of NO_ERROR after sending a complete response".`

We do see the response HEADERS but not the body, perhaps it was dropped for whatever reason. We'll attempt to get a backtrace of the problematic response and see whether we're failing e.g. too early and should have processed the response (onData) body as well but didn't.

**Comment by ppadevski-vmw** [comment]
After a more thorough look I think that this may be an actual bug in envoy.

What puzzled me at first was the following pcap:

Transmission Control Protocol, Src Port: 1080, Dst Port: 60986, Seq: 552713, Ack: 1251776, Len: 9147
```
HyperText Transfer Protocol 2
    Stream: HEADERS, Stream ID: 3423, Length 61, 200 OK
HyperText Transfer Protocol 2
    Stream: DATA, Stream ID: 3423, Length 1512 (partial entity body)
HyperText Transfer Protocol 2
    Stream: DATA, Stream ID: 3423, Length 0
    eXtensible Markup Language
HyperText Transfer Protocol 2
    Stream: RST_STREAM, Stream ID: 3423, Length 4
HyperText Transfer Protocol 2
    Stream: HEADERS, Stream ID: 3439, Length 61, 200 OK
HyperText Transfer Protocol 2
    Stream: DATA, Stream ID: 3439, Length 1512 (partial entity body)
HyperText Transfer Protocol 2
    Stream: DATA, Stream ID: 3439, Length 0
    eXtensible Markup Language
HyperText Transfer Protocol 2
    Stream: RST_STREAM, Stream ID: 3439, Length 4
HyperText Transfer Protocol 2
    Stream: HEADERS, Stream ID: 3441, Length 61, 200 OK
HyperText Transfer Protocol 2
    Stream: DATA, Stream ID: 3441, Length 1512 (partial entity body)
HyperText Transfer Protocol 2
    Stream: DATA, Stream ID: 3441, Length 0
    eXtensible Markup Language
HyperText Transfer Protocol 2
    Stream: RST_STREAM, Stream ID: 3441, Length 4
HyperText Transfer Protocol 2
    Stream: HEADERS, Stream ID: 3437, Length 61, 200 OK
HyperText Transfer Protocol 2
    Stream: DATA, Stream ID: 3437, Length 1512 (partial entity body)
HyperText Transfer Protocol 2
    Stream: DATA, Stream ID: 3437, Length 0
    eXtensible Markup Language
HyperText Transfer Protocol 2
    Stream: RST_STREAM, Stream ID: 3437, Length 4
HyperText Transfer Protocol 2
    Stream: HEADERS, Stream ID: 3431, Length 61, 200 OK
HyperText Transfer Protocol 2
    Stream: DATA, Stream ID: 3431, Length 1512 (partial entity body)
HyperText Transfer Protocol 2
    Stream: DATA, Stream ID: 3431, Length 0
    eXtensible Markup Language
HyperText Transfer Protocol 2
    Stream: RST_STREAM, Stream ID: 3431, Length 4
HyperText Transfer Protocol 2
    Stream: HEADERS, Stream ID: 3429, Length 61, 200 OK
HyperText Transfer Protocol 2
    Stream: DATA, Stream ID: 3429, Length 824 (partial entity body)
HyperText Transfer Protocol 2
    Stream: RST_STREAM, Stream ID: 3429, Length 4
HyperText Transfer Protocol 2
    Stream: HEADERS, Stream ID: 3457, Length 61, 200 OK
HyperText Transfer Protocol 2
    Stream: RST_STREAM, Stream ID: 3457, Length 4
HyperText Transfer Protocol 2
    Stream: HEADERS, Stream ID: 3455, Length 61, 200 OK
HyperText Transfer Protocol 2
    Stream: RST_STREAM, Stream ID: 3455, Length 4
```

Notice that we more or less send the exact same response (1512 bytes) and then suddenly we send a partial response (824 bytes) and then everything after that is completely busted (including the stream with the partial response). Basically we're good and then we get lots of RST_STREAMs.

For the record all of the busted response go through here:

```
  // If we submit a reset, nghttp2 will cancel outbound frames that have not yet been sent.
  // We want these frames to go out so we defer the reset until we send all of the frames that
  // end the local stream. However, if we're resetting the stream due to
  // overload, we should reset the stream as soon as possible to free used
  // resources.
  if (useDeferredReset() && local_end_stream_ && !local_end_stream_sent_ &&
      reason != StreamResetReason::OverloadManager) {
    ASSERT(parent_.getStreamUnchecked(stream_id_) != nullptr);
    parent_.pending_deferred_reset_streams_.emplace(stream_id_, this);
    deferred_reset_ = reason;
    ENVOY_CONN_LOG(error, "deferred reset stream", parent_.connection_); // <<<<<<<<<<<<<<<<<<
  } else {
    resetStreamWorker(reason);
  }
```

So that is, local_end_stream_ is true and local_end_stream_sent_ is false, that is we didn't send the data frame(s) with End Stream.

The interesting part is why that happened and my best guess so far is flow control. I am very certain that the client session window was exhausted (the client reads responses a little bit slower and its window was set to 65k).

If we assume that we did in fact hit flow control my best guess is that the issue is here:

```
Status ConnectionImpl::sendPendingFrames() {
  if (dispatching_ || connection_.state() == Network::Connection::State::Closed) {
    return okStatus();
  }

  const int rc = adapter_->Send(); /// <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
  if (rc != 0) {
    ASSERT(rc == NGHTTP2_ERR_CALLBACK_FAILURE);
    return codecProtocolError(nghttp2_strerror(rc));
  }

  // See ConnectionImpl::StreamImpl::resetStream() for why we do this. This is an uncommon event,
  // so iterating through every stream to find the ones that have a deferred reset is not a big
  // deal. Furthermore, queueing a reset frame does not actually invoke the close stream callback.
  // This is only done when the reset frame is sent. Thus, it's safe to work directly with the
  // stream map.
  // NOTE: The way we handle deferred reset is essentially best effort. If we intend to do a
  //       deferred reset, we try to finish the stream, including writing any pending data frames.
  //       If we cannot do this (potentially due to not enough window), we just reset the stream.
  //       In general this behavior occurs only when we are trying to send immediate error messages
  //       to short circuit requests. In the best effort case, we complete the stream before
  //       resetting. In other cases, we just do the reset now which will blow away pending data
  //       frames and release any memory associated with the stream.
  if (!pending_deferred_reset_streams_.empty()) {
    while (!pending_deferred_reset_streams_.empty()) {
      auto it = pending_deferred_reset_streams_.begin();
      auto* stream = it->second;
      // Sanity check: the stream's id matches the map key.
      ASSERT(it->first == stream->stream_id_);
      pending_deferred_reset_streams_.erase(it);
      ASSERT(stream->deferred_reset_);
      stream->resetStreamWorker(stream->deferred_reset_.value());
    }
    RETURN_IF_ERROR(sendPendingFrames());
  }
```

Basically the Send() + (deferred) RST recipe above would only work if Send() actually does something. If it does nothing due to flow control, we'll basically end up sending only RSTs.

See e.g.
https://github.com/google/quiche/blob/main/quiche/http2/adapter/nghttp2_adapter.cc#L150
https://nghttp2.org/documentation/nghttp2_session_send.html
https://github.com/nghttp2/nghttp2/blob/b7f98539687d4857ff34e9b8306bbe5b722e2916/lib/nghttp2_session.c#L3527

In any case I am not an expert in these areas but I am pretty certain that we did in fact hit flow control. I have a change that increases the window size of the client to 1mb which may fix the immediate issue but I still believe that something can be done in envoy to improve this behaviour.

**Comment by wbpcode** [comment]
cc @htuch for possible input for this http2-related problem.

**Comment by ppadevski-vmw** [comment]
In other words is it possible to do e.g.

```
  if (!pending_deferred_reset_streams_.empty()) {
    while (!pending_deferred_reset_streams_.empty()) {
      auto it = pending_deferred_reset_streams_.begin();
      auto* stream = it->second;
      // Sanity check: the stream's id matches the map key.
      ASSERT(it->first == stream->stream_id_);
      pending_deferred_reset_streams_.erase(it);
      ASSERT(stream->deferred_reset_);
      if (stream->local_end_stream_sent_) {
        // we're good
        stream->resetStreamWorker(stream->deferred_reset_.value());
      } else {
        // add to another map or reinsert (be mindful of recursive sendPendingFrames() below)
        // wait for NGHTTP2_WINDOW_UPDATE from onFrameReceived() and then do another sendPendingFrames() if pending_deferred_reset_streams_ is not empty
      }
    }
    RETURN_IF_ERROR(sendPendingFrames());
  }
```

**Comment by htuch** [comment]
@yanavlasov

**Comment by ppadevski-vmw** [comment]
I will probably use the proposed patch for some time and will report back if we see any issues with it.
